lib/ansible/utils/encrypt.py

```diff
@@ -148,9 +159,15 @@ def _clean_salt(self, salt):
         if not salt:
             return None
         elif issubclass(self.crypt_algo, HasRawSalt):
-            return to_bytes(salt, encoding='ascii', errors='strict')
+            ret = to_bytes(salt, encoding='ascii', errors='strict')
         else:
-            return to_text(salt, encoding='ascii', errors='strict')
+            ret = to_text(salt, encoding='ascii', errors='strict')
+
+        # Ensure the salt has the correct padding
+        if self.algorithm == 'bcrypt':
+            ret = bcrypt64.repair_unused(ret)
```

Comment:
Should we be worried that the current [passlib docs](https://passlib.readthedocs.io/en/stable/lib/passlib.utils.binary.html) for the module with this method has a large warning at the top?

```
Warning: This module is primarily used as an internal support module. Its interface has not been finalized yet, and may be changed somewhat between major releases of Passlib, as the internal code is cleaned up and simplified.
```

Reply:
I've gone over this all again, and the original warning:

> encountered a bcrypt salt with incorrectly set padding bits; you may want to use bcrypt.normhash() to fix this; this will be an error under Passlib 2.0

...indicates that for now passlib will allow itself to generate an invalid hash, when given in salt with incorrect padding bits, but in the future it will not.

So I think overall the implementation is correct here, because we want to give the user a valid hash, whether they gave us a salt, or we generated the salt.

However, the only method available within passlib to fix the salt is this `repair_unused`.  I could re-implement it, although with likely being less exact, or rely on the potential that this `bcrypt64` module could change.

The `bcrypt64` module within `passlib` appears to have been stable enough since at least 1.6.5 when it existed in a different location.

So to get back to your question, we should probably be cautious, and adding a test will help us identify issues with any change that is made.

Reply:
Ok, I just pushed a test which should cover it.  Let me know if it addresses your concern.